### PR TITLE
[Scenic 3] Add the Parser & Compiler Guide

### PR DESCRIPTION
**Pull request by shumbo**
This PR adds `compiler.rst` that describes the new parser & compiler architecture. It also updates the flow of compilation in `internals.rst`.

### CODE REVIEW

File: docs/compiler.rst [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+
+.. figure:: /images/parser_architecture.png
+  :alt: Scenic Parser & Compiler Architecture
+  :figclass: align-center
+
+Scenic AST
+~~~~~~~~~~
+
+
+Scenic AST is an abstract syntax tree for representing Scenic programs.
```
Comment (Eric-Vin) [inline docs/compiler.rst:15]: [A] Scenic AST

File: docs/compiler.rst [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+It is a superset of Python AST and includes nodes for Scenic-specific
+language constructs.
+
+``syntax/ast.py`` defines all Scenic-specific AST nodes. Note that all
+Scenic AST nodes must extend the ``AST`` class defined on the same file.
```
Comment (Eric-Vin) [inline docs/compiler.rst:20]: ... defined [in] the same file.

File: docs/compiler.rst [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+
+AST nodes should include fields to store objects. To add fields, add a
+parameter to the initializer and define fields by assigning values to
+``self``.
+
+When adding fields, be sure to update the ``_fields`` and
+``__match_args__`` fields. ``__fields`` lists the name of the fields on
```
Comment (Eric-Vin) [inline docs/compiler.rst:27]: ... the name[s] of the fields [in] ...

File: docs/compiler.rst [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+the AST node and is used by the AST module to traverse the tree, fill
+the missing information, etc. ``__match_args__`` is used by the test
```
Comment (Eric-Vin) [inline docs/compiler.rst:29]: ... fill [in] missing information ...

File: docs/compiler.rst [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+Scenic Parser
+~~~~~~~~~~~~~
+
+Scenic Parser takes a Scenic source code and outputs the corresponding
+abstract syntax tree. It is generated from the grammar file using Pegen.
+
+When you modify ``scenic.gram``, you need to generate the parser again.
+Call ``make`` at the project root to do so. Pytest is configured to
+update the parser before test execution.
+
```
Comment (Eric-Vin) [inline docs/compiler.rst:51]: I believe that the makefile won't work for Windows users outside of WSL, so maybe we should add a note here on how to run the command manually for them?

Comment (Eric-Vin) [reply docs/compiler.rst:51]: Also does Pytest update the parser before test execution (if changes have been made), or just generate it if it doesn't already exist?

File: docs/compiler.rst [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+``test_parser.py`` includes parser tests and ensures that the parser
+generates the desired AST.
+
+Scenic Compiler
+~~~~~~~~~~~~~~~
+
+Scenic Compiler is a Scenic AST-to-Python AST compiler. The generated
+Python AST would be passed to the Python interpreter for execution.
```
Comment (Eric-Vin) [inline docs/compiler.rst:59]: ... Python AST [can] be passed ...

File: docs/compiler.rst [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+
+Internally, Scenic Compiler is a subclass of ``NodeTransformer``. It
+must define visitors for each Scenic AST node and return corresponding
+Python AST nodes.
+
+Tutorial: Adding New Syntax
+---------------------------
+
+In order to add new syntax, you want to follow the following steps:
```
Comment (Eric-Vin) [inline docs/compiler.rst:68]: ... new syntax [you'll want to do the following]:

File: docs/compiler.rst [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+   syntax tree.
+
+Step 2: Add Grammar
+~~~~~~~~~~~~~~~~~~~
+
+.. note::
+    
+    The grammar described here is simplified for the sake of brevity and is not the actual grammar used in the Scenic language.
+
+The next step is to add a grammar to the ``scenic.gram`` file.
```
Comment (Eric-Vin) [inline docs/compiler.rst:123]: ... is to [update] the ``scenic.gram`` file.

File: docs/compiler.rst [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+
+::
+
+   scenic_implication (memo):
+       | invalid_scenic_implication
+       | a=disjunction 'implies' b=disjunction { s.ImpliesOp(a, b, LOCATIONS) }
+       | disjunction
+
+In the grammar, the precedence and associativity is given by using
```
Comment (Eric-Vin) [inline docs/compiler.rst:132]: ...the grammar, [] precedence and...

File: docs/compiler.rst [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+separate rules for each precedence. Since implication should bind less tightly than ``or``,
+it takes two ``disjunction``\ s for its operands.
+
+All Scenic grammar rules should be prefixed with ``scenic_`` so we can
+easily distinguish rules added to the original Python grammar.
```
Comment (Eric-Vin) [inline docs/compiler.rst:137]: ... easily distinguish [Scenic specific rules from those in] the original Python grammar.

File: docs/compiler.rst [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+
+The production rule is defined inside the brackets. ``s`` refers to the
+Scenic AST module and we can write the plain Python to construct the AST
+node. ``LOCATIONS`` will be replaced with a set of named arguments to
+express source code locations.
+
+The ``implies`` operator is unique in that it takes exactly two
+operands. Since the associativity of ``implies`` operator is not clear,
+developers must specify the associativity by explicitly wrapping one of
+the operand with parenthesis.
```
Comment (Eric-Vin) [inline docs/compiler.rst:147]: ... the operand[s] with ...

File: docs/compiler.rst [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+::
+
+   invalid_scenic_implication[NoReturn]:
+       | a=disjunction 'implies' implication {
+           self.raise_syntax_error_known_location(
+               f"`implies` must take exactly two operands", a
+           )
+        }
+
+The ``invalid_scenic_implication`` rule looks for implication with more
```
Comment (Eric-Vin) [inline docs/compiler.rst:163]: ... looks for [an] implication ...

File: docs/internals.rst [hunk 13]
```diff
@@ -29,45 +29,25 @@ The ``scenic`` module itself provides the top-level API for using Scenic: see :d
...
+If you want to make a change to the language, also see :doc:`compiler`.
 
-Phase 3: Python Parser
+Phase 1: Scenic Parser
 ~~~~~~~~~~~~~~~~~~~~~~
-In this phase we parse the token stream resulting from Phase 2 using the Python parser.
-If the tokens do not parse properly, they come from a malformed Scenic program, and an error will be raised here.
-Unfortunately, since the tokens being parsed are not the original Scenic tokens and likely not semantically-valid Python either, the error messages can be confusing.
+In this phase the program is parsed using the Scenic parser. The parser is generated from a PEG grammar (``scenic.gram``) using `the Pegen parser generator <https://we-like-parsers.github.io/pegen/index.html>`_.
+The parser generates an abstract syntax tree (Scenic AST) for the program. Scenic AST is a superset of Python AST defined in ``ast.py`` and has additional nodes for representing Scenic-specific constructs.
```
Comment (Eric-Vin) [inline docs/internals.rst:40]: Maybe we should add a link here to the new guide, for people who want more details?

**Comment by Eric-Vin** [comment]
This guide is great, thanks Shun! I left some comments with very minor tweaks to wording, but feel free to ignore any you don't think sound better. The only significant thing I think we should add is more specific instructions for Windows users who can't use the Makefile.

**Review by Eric-Vin — COMMENTED** [review]
This guide is great, thanks @shumbo! I left some comments with very minor tweaks to wording, but feel free to ignore any you don't think sound better. The only significant thing I think we should add is more specific instructions for Windows users who can't use the Makefile.

File: docs/compiler.rst (hunk 2 above)
Comment (shumbo) [reply docs/compiler.rst]: fixed 4cf64494a4ef4340c82445ac97db27fe9d1dd05b

File: docs/compiler.rst (hunk 3 above)
Comment (shumbo) [reply docs/compiler.rst]: fixed 4cf64494a4ef4340c82445ac97db27fe9d1dd05b

File: docs/compiler.rst (hunk 6 above)
Comment (shumbo) [reply docs/compiler.rst]: fixed 4cf64494a4ef4340c82445ac97db27fe9d1dd05b

File: docs/compiler.rst (hunk 7 above)
Comment (shumbo) [reply docs/compiler.rst]: fixed 4cf64494a4ef4340c82445ac97db27fe9d1dd05b

File: docs/compiler.rst (hunk 9 above)
Comment (shumbo) [reply docs/compiler.rst]: fixed 4cf64494a4ef4340c82445ac97db27fe9d1dd05b

File: docs/compiler.rst (hunk 11 above)
Comment (shumbo) [reply docs/compiler.rst]: fixed 4cf64494a4ef4340c82445ac97db27fe9d1dd05b

File: docs/compiler.rst (hunk 12 above)
Comment (shumbo) [reply docs/compiler.rst]: fixed 4cf64494a4ef4340c82445ac97db27fe9d1dd05b

File: docs/compiler.rst (hunk 10 above)
Comment (shumbo) [reply docs/compiler.rst]: fixed https://github.com/BerkeleyLearnVerify/Scenic/commit/4cf64494a4ef4340c82445ac97db27fe9d1dd05b

**Review by Eric-Vin — APPROVED** [review]
Looks great to me!

**Comment by dfremont** [comment]
Looks good to me too -- thanks @shumbo! I thought the Internals page was getting long, so I reorganized things a bit and revised the text as I went. I also took the opportunity to add the Pegen lexer to highlight the snippets in your tutorial, as well as the full grammar. If it all looks good to you I will merge.

**Review by dfremont — APPROVED** [review]

**Pull request merged**